File: json-schema/research/format_hostname/runner.mjs
// JS runner for the `hostname` format conformance corpus (TypeScript target).
//
// Implements the PINNED generator-owned check:
//   1. module-level RegExp with the `u` flag, fully anchored ^...$
//   2. a total-length guard (1..=253 CODE POINTS) OUTSIDE the regex.
//      IMPORTANT: use [...v].length (code points), not v.length (UTF-16 units),
//      so the length agrees with the other targets on astral input. (Hostnames
//      are ASCII so this only matters for the non-ASCII reject rows, where the
//      regex already rejects; kept correct on principle.)
// Verdict = (regex matches) AND (length in range).
//
// JS `$` (no `m` flag) matches end-of-input only -- the portable choice.
//
// Emits JSON Lines: {"id","engine":"js","valid","regex","len_ok"}
// Run: node runner.mjs [corpus.json]
import { readFileSync } from "node:fs";

const HOST_RE =
  /^[A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?(?:\.[A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?)*$/u;
const MAX_TOTAL_LEN = 253;

const path = process.argv[2] ?? "corpus.json";
const corpus = JSON.parse(readFileSync(path, "utf8"));

for (const k of corpus.cases) {
  const n = [...k.instance].length; // code points
  const lenOk = n >= 1 && n <= MAX_TOTAL_LEN;
  const regex = HOST_RE.test(k.instance);
  process.stdout.write(
    JSON.stringify({
      id: k.id,
      engine: "js",
      valid: regex && lenOk,
      regex,
      len_ok: lenOk,
    }) + "\n",
  );
}

File: json-schema/research/format_hostname/runner.go
// Go runner for the `hostname` format conformance corpus.
//
// Implements the PINNED generator-owned check for the Go target:
//   1. package-level compiled regex (RE2, ASCII), fully anchored ^...$
//   2. a total-length guard (1..=253 code points) OUTSIDE the regex
//      (RE2 has no lookahead for a whole-input length assertion)
// The verdict is (regex matches) AND (length in range).
//
// Reads corpus.json (argv[1] or ./corpus.json) and emits JSON Lines to stdout:
//   {"id","engine":"go","valid":bool,"regex":bool,"len_ok":bool}
//
// Run: go run runner.go [corpus.json]
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"regexp"
	"unicode/utf8"
)

// PINNED hostname regex: labels of [A-Za-z0-9-], 1-63 chars, no leading/
// trailing hyphen, separated by '.', at least one label. Fully anchored.
// Go's `$` matches end-of-input only (no trailing-\n exception) -- the
// portable choice.
var hostRe = regexp.MustCompile(`^[A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?(?:\.[A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?)*$`)

const maxTotalLen = 253

type kase struct {
	ID       string `json:"id"`
	Instance string `json:"instance"`
	Valid    bool   `json:"valid"`
}
type corpus struct {
	Cases []kase `json:"cases"`
}
type result struct {
	ID     string `json:"id"`
	Engine string `json:"engine"`
	Valid  bool   `json:"valid"`
	Regex  bool   `json:"regex"`
	LenOK  bool   `json:"len_ok"`
}

func main() {
	path := "corpus.json"
	if len(os.Args) > 1 {
		path = os.Args[1]
	}
	data, err := os.ReadFile(path)
	if err != nil {
		fmt.Fprintln(os.Stderr, "read error:", err)
		os.Exit(1)
	}
	var c corpus
	if err := json.Unmarshal(data, &c); err != nil {
		fmt.Fprintln(os.Stderr, "parse error:", err)
		os.Exit(1)
	}
	enc := json.NewEncoder(os.Stdout)
	for _, k := range c.Cases {
		n := utf8.RuneCountInString(k.Instance)
		lenOK := n >= 1 && n <= maxTotalLen
		rx := hostRe.MatchString(k.Instance)
		enc.Encode(result{
			ID: k.ID, Engine: "go",
			Regex: rx, LenOK: lenOK, Valid: rx && lenOK,
		})
	}
}
